File: 2022/03/1.py
from input import input


def score(intersect):
    if ord(intersect) > ord('Z'):
        return ord(intersect) - ord('a') + 1
    else:
        return ord(intersect) - ord('A') + 27


sum = 0

for items in input.split('\n'):
    p1 = items[:len(items)//2]
    p2 = items[len(items)//2:]
    intersect = ""
    for i in p1:
        if i in p2 and not i in intersect:
            intersect += i
    sum += score(intersect)

print(sum)

File: 2023/14/1.rs
use std::fs;

const EMPTY: char = '.';
const MOVE: char = 'O';

type Map = Vec<Vec<char>>;

fn main() {
    let file_path = "./input";
    let input = fs::read_to_string(file_path).expect("Should have been able to read the file");

    let mut map: Map = input.lines().map(|l| l.chars().collect()).collect();

    north(&mut map);

    println!(
        "{}",
        map.iter()
            .rev()
            .enumerate()
            .map(|(score, col)| (score + 1) * col.iter().filter(|&&c| c == MOVE).count())
            .sum::<usize>()
    )
}

fn north(map: &mut Map) {
    let w = map[0].len();
    let h = map.len();

    for col in 0..w {
        let mut prev = map[0][col];
        let mut move_to: usize = 0;

        for row in 1..h {
            let mut curr = map[row][col];
            if prev != EMPTY {
                move_to = row;
            }
            if curr == MOVE {
                map[row][col] = EMPTY;
                map[move_to][col] = MOVE;
                curr = map[row][col];
                move_to += 1;
            }

            prev = curr;
        }
    }
}
